quantization: align observer dtype with reference model spec

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#85345 quantization: align observer dtype with reference model spec**

Summary:

Before this PR, the `dtype` attribute of observers was not clearly
defined.  It originally meant `interface_dtype` in the eager mode
workflow, which is how the codebase before this PR is using it.

In the new reference model spec, `dtype` attribute of an observer
represents the `dtype` value which needs to be passed into a `quantize`
function in the reference model spec. This PR aligns the codebase
to this definition of dtype.  In detail:
1. change util functions to interpret `dtype` using the reference model definition
2. change `prepare` to interpret `dtype` using the reference model definition
3. change observers for dynamic quantization to interpret `dtype` using the reference
   model definition.

A future PR (left out of this one to keep LOC small) will deprecate the
`compute_dtype` field and instead expose `is_dynamic` on observers.
"

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Differential Revision: [D39675209](https://our.internmc.facebook.com/intern/diff/D39675209)

vkuzo committed:
quantization: align observer dtype with reference model spec

Summary:

Before this PR, the `dtype` attribute of observers was not clearly
defined.  It originally meant `interface_dtype` in the eager mode
workflow, which is how the codebase before this PR is using it.

In the new reference model spec, `dtype` attribute of an observer
represents the `dtype` value which needs to be passed into a `quantize`
function in the reference model spec. This PR aligns the codebase
to this definition of dtype.  In detail:
1. change util functions to interpret `dtype` using the reference model definition
2. change `prepare` to interpret `dtype` using the reference model definition
3. change observers for dynamic quantization to interpret `dtype` using the reference
   model definition.

A future PR (left out of this one to keep LOC small) will deprecate the
`compute_dtype` field and instead expose `is_dynamic` on observers.
"

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

[ghstack-poisoned]

z-a-f commented on torch/ao/quantization/utils.py:
```diff
@@ -183,11 +183,22 @@ def weight_dtype(qconfig):
+    activation = qconfig.activation()
+    if hasattr(activation, 'compute_dtype'):
+        return True
+    return None
```
Can we return `False` instead? To me it is harder to understand and there is a subtle difference between`if is_dynamic(...) is None` vs `if is_dynamic(...)`

vkuzo replied:
this matches existing backend_config_dict code, let's split cleaning this up in a separate PR

vkuzo commented on torch/ao/quantization/observer.py:
```diff
@@ -1322,6 +1333,7 @@ def __init__(
         self.dtype = dtype
         self.custom_op = custom_op_name
         # used for configuration of computation type for dynamic quantization
+        # TODO(future PR): replace this with `is_dynamic`
```
it would be more involved, leaving for separate PR to keep them smaller

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -174,35 +174,37 @@ def is_input_arg_dtype_supported_by_backend(
-                node_name_to_target_dtype[node.name]["input_activation_dtype"] == input_activation_dtype
+            qconfig_dtype, qconfig_is_dynamic = None, None
+        # TODO(future PR): remove the cast to bool below after figuring
+        # out why backend_config has is_dynamic set to None in some cases.
```
probably because it's initialized to None, I feel we can initialize it to False

vkuzo replied:
tried that, didn't fully solve the issue.  Let's do in a separate PR.

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1539,7 +1551,7 @@ def load_observer_state_dict(mod, obs_dict):
 
 default_dynamic_quant_observer = PlaceholderObserver.with_args(
-    dtype=torch.float, compute_dtype=torch.quint8
+    dtype=torch.quint8, compute_dtype=torch.quint8
```
do we need `is_dynamic` here?

jerryzh168 replied:
oh, saw this in the summary, so now we are still using `dtype` and `compute_dtype`, we'll use `dtype` and `is_dynamic` in a future PR

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -157,7 +157,7 @@ def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
 """
 
-float16_dynamic_qconfig = QConfig(activation=PlaceholderObserver.with_args(dtype=torch.float32, compute_dtype=torch.float16),
+float16_dynamic_qconfig = QConfig(activation=PlaceholderObserver.with_args(dtype=torch.float16, compute_dtype=torch.float16),
```
do we need `is_dynamic` here?

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -183,11 +183,22 @@ def weight_dtype(qconfig):
-    return activation_dtype(qconfig) in [torch.quint8, torch.qint8, torch.float16]
+    return (
+        activation_dtype(qconfig) in [torch.quint8, torch.qint8, torch.float16]
+        and is_dynamic(qconfig) is None
```
maybe `not is_dynamic(qconfig)`?

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -183,11 +183,22 @@ def weight_dtype(qconfig):
     return weight.dtype
 
+# Note: the type is Optional[bool] to match backend_config_dict
+def is_dynamic(qconfig) -> Optional[bool]:
```
this is checking activation only, should we change the name

jerryzh168 reviewed (approved):
lg, thanks

jerryzh168 reviewed (commented):
I'm wondering if it would make more sense to change this directly to the end state though, since currently looks like we are in a middle state that does not align with the old or new design

vkuzo committed:
Update on "quantization: align observer dtype with reference model spec"


Summary:

Before this PR, the `dtype` attribute of observers was not clearly
defined.  It originally meant `interface_dtype` in the eager mode
workflow, which is how the codebase before this PR is using it.

In the new reference model spec, `dtype` attribute of an observer
represents the `dtype` value which needs to be passed into a `quantize`
function in the reference model spec. This PR aligns the codebase
to this definition of dtype.  In detail:
1. change util functions to interpret `dtype` using the reference model definition
2. change `prepare` to interpret `dtype` using the reference model definition
3. change observers for dynamic quantization to interpret `dtype` using the reference
   model definition.

A future PR (left out of this one to keep LOC small) will deprecate the
`compute_dtype` field and instead expose `is_dynamic` on observers.
"

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

[ghstack-poisoned]

vkuzo commented:
> since currently looks like we are in a middle state that does not align with the old or new design

This was also true before this PR, because we had backend_config_dict use `dtype` as "argument to quantize", but dynamic quantization observers use `dtype` as "interface dtype".  So, this PR is just moving from a messy state to a less messy state, it's not moving from a clean state to a messy state.

vkuzo committed:
Update on "quantization: align observer dtype with reference model spec"


Summary:

Before this PR, the `dtype` attribute of observers was not clearly
defined.  It originally meant `interface_dtype` in the eager mode
workflow, which is how the codebase before this PR is using it.

In the new reference model spec, `dtype` attribute of an observer
represents the `dtype` value which needs to be passed into a `quantize`
function in the reference model spec. This PR aligns the codebase
to this definition of dtype.  In detail:
1. change util functions to interpret `dtype` using the reference model definition
2. change `prepare` to interpret `dtype` using the reference model definition
3. change observers for dynamic quantization to interpret `dtype` using the reference
   model definition.

A future PR (left out of this one to keep LOC small) will deprecate the
`compute_dtype` field and instead expose `is_dynamic` on observers.
"

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

[ghstack-poisoned]

vkuzo committed:
Update on "quantization: align observer dtype with reference model spec"


Summary:

Before this PR, the `dtype` attribute of observers was not clearly
defined.  It originally meant `interface_dtype` in the eager mode
workflow, which is how the codebase before this PR is using it.

In the new reference model spec, `dtype` attribute of an observer
represents the `dtype` value which needs to be passed into a `quantize`
function in the reference model spec. This PR aligns the codebase
to this definition of dtype.  In detail:
1. change util functions to interpret `dtype` using the reference model definition
2. change `prepare` to interpret `dtype` using the reference model definition
3. change observers for dynamic quantization to interpret `dtype` using the reference
   model definition.

A future PR (left out of this one to keep LOC small) will deprecate the
`compute_dtype` field and instead expose `is_dynamic` on observers.
"

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39675209).
